File: sol/solution/0001-0100/0016. 3Sum Closest/Solution.scala
object Solution {
    def threeSumClosest(nums: Array[Int], target: Int): Int = {
        val sorted = nums.sorted
        val n = nums.length
        var closestSum = sorted(0) + sorted(1) + sorted(2)
        var found = false
        var result = closestSum
        var i = 0
        while (i < n - 2 && !found) {
            var left = i + 1
            var right = n - 1
            while (left < right && !found) {
                val currSum = sorted(i) + sorted(left) + sorted(right)
                if (currSum == target) {
                    result = currSum
                    found = true
                } else {
                    if (Math.abs(currSum - target) < Math.abs(closestSum - target)) {
                        closestSum = currSum
                    }
                    if (currSum < target) left += 1
                    else right -= 1
                }
            }
            i += 1
        }
        if (found) result else closestSum
    }
}

File: sol/solution/0001-0100/0016. 3Sum Closest/Solution.rs
impl Solution {
    pub fn three_sum_closest(nums: Vec<i32>, target: i32) -> i32 {
        let mut nums = nums;
        nums.sort();
        let n = nums.len();
        let mut closest_sum = nums[0] + nums[1] + nums[2];
        for i in 0..n-2 {
            let mut left = i + 1;
            let mut right = n - 1;
            while left < right {
                let curr_sum = nums[i] + nums[left] + nums[right];
                if curr_sum == target {
                    return curr_sum;
                }
                if (curr_sum - target).abs() < (closest_sum - target).abs() {
                    closest_sum = curr_sum;
                }
                if curr_sum < target {
                    left += 1;
                } else {
                    right -= 1;
                }
            }
        }
        closest_sum
    }
}